File: Array/huangxinyu/c/MyQueue.cpp
//
// Created by hasee on 2020/6/17.
//
#include <stack>

class MyQueue {
public:
    /** Initialize your data structure here. */
    MyQueue() {

    }

    /** Push element x to the back of queue. */
    void push(int x) {
        st1.push(x);
    }

    /** Removes the element from in front of queue and returns that element. */
    int pop() {
        int top = peek();
        st2.pop();
        return top;
    }

    /** Get the front element. */
    int peek() {
        if (st2.empty()) {
            while (!st1.empty()) {
                st2.push(st1.top());
                st1.pop();
            }
        }
        return st2.top();
    }

    /** Returns whether the queue is empty. */
    bool empty() {
        return st1.empty() && st2.empty();
    }

private:
    std::stack<int> st1, st2;
};

File: Array/huangxinyu/kotlin/src/MyQueue.kt
import java.util.*

class MyQueue {
    val stack1: Stack<Int> = Stack<Int>()
    val stack2: Stack<Int> = Stack<Int>()
    fun push(x: Int) {
        stack1.push(x)
    }

    fun peek(): Int {
        if (stack2.empty()) {
            while (!stack1.empty()) {
                val peek = stack1.peek()
                stack2.push(peek)
                stack1.pop()
            }
        }
        return stack2.peek()
    }

    fun pop(): Int {
        val top = peek()
        stack2.pop()
        return top
    }

    fun empty(): Boolean {
        return stack1.empty() && stack2.empty()
    }
}